File: solutions/977. Squares of a Sorted Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        #soln 2 - Leetcode two pointers, Time O(N), Space O(N)
        l = len(nums)
        ans = [0] * l
        negptr, posptr= 0, l-1
        
        for i in range(l-1, -1, -1):
            if abs(nums[negptr]) < abs(nums[posptr]):
                sq = nums[posptr]
                posptr -= 1
            else:
                sq = nums[negptr]
                negptr += 1
            ans[i] = sq*sq
        return ans
        
#         #soln 1 - too complicated, Time O(N), Space O(N)
#         l = len(nums)
#         ans = [None] * l
#         i = 0
#         k = j = l-1
        
#         while i < len(nums) and nums[i] < 0:
#             while abs(nums[i]) <= nums[j]:
#                 ans[k] = nums[j] * nums[j]
#                 j -= 1
#                 k -= 1
#             ans[k] = nums[i] * nums[i] 
#             k -= 1 
#             i += 1
#         ans[:k+1] = [n*n for n in nums[i:i+k+1]]  
#         return ans
        
        #soln 0 - trivial
        #return sorted([i*i for i in nums])

File: solutions/977. Squares of a Sorted Array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        // Two pointers, time O(N), space O(N)
        int numsLen = nums.size();
        int r = numsLen-1, l = 0;
        vector<int> ans(numsLen);
        for (int i = numsLen-1; i > -1; i--) {
            if (abs(nums[l]) > abs(nums[r])) {
                ans[i] = nums[l]*nums[l];
                l++;
            }
            else {
                ans[i] = nums[r]*nums[r--];
                r--;
            }
        }
        return ans;
        
        // // trivial, time O(NlogN), space O(N)
        // vector<int> ans;
        // for (auto num : nums) {
        //     ans.push_back(num * num);
        // }
        // sort(begin(ans), end(ans));
        // return ans;
    }
};
